File: js/ds/Graph.js
// https://www.geeksforgeeks.org/implementation-graph-javascript/
class Graph {
  constructor(numVertices) {
    this.numVertices = numVertices;
    this.adjList = new Map();
  }

  addVertex(v) {
    this.adjList.set(v, []);
  }

  addEdge(v, w) {
    this.adjList.get(v).push(w); // vertex w denotes edge between v and w

    // for undirected graph, add edge between w to v as well
    this.adjList.get(w).push(v);
  }

  printGraph() {
    this.adjList.forEach((edges, vertex) => {
      console.log(`${vertex} => ${edges}`);
    });
  }

  bfs(root) {
    let visited = {};
    let queue = [];

    // initialise the queue with root vertex
    visited[root] = true;
    queue.push(root);

    while (queue.length > 0) {
      let current = queue.shift();
      // for now just print out
      console.log(current);

      // get current's neighbours
      let neighbours = this.adjList.get(current);
      // push them to the queue
      neighbours.forEach((neighbour) => {
        if (!visited[neighbour]) {
          visited[neighbour] = true;
          queue.push(neighbour);
        }
      });
    }
  }

  printLevelsUsingBFS(root) {
    let visited = {};
    let queue = [];
    let levels = [];

    // initialise the queue with root vertex
    visited[root] = true;
    queue.push(root);
    levels[root] = 0;

    while (queue.length > 0) {
      let current = queue.shift();
      // for now just print out
      console.log(`level: ${levels[current]} - ${current}`);

      // get current's neighbours
      let neighbours = this.adjList.get(current);
      // push them to the queue
      neighbours.forEach((neighbour) => {
        if (!visited[neighbour]) {
          visited[neighbour] = true;
          queue.push(neighbour);

          // level of neighbour = level of current + 1
          levels[neighbour] = levels[current] + 1;
        }
      });
    }
  }

  dfs(root) {
    let visited = {};
    this.dfTraverse(root, visited);
  }

  dfTraverse(vert, visited) {
    visited[vert] = true;
    console.log(vert);

    let neighbours = this.adjList.get(vert);
    neighbours.forEach((neighbour) => {
      if (!visited[neighbour]) {
        this.dfTraverse(neighbour, visited);
      }
    });
  }
}

const graph = new Graph();

const VERTICIES = ['A', 'B', 'C', 'D', 'E', 'F'];
VERTICIES.forEach((v) => graph.addVertex(v));

graph.addEdge('A', 'B');
graph.addEdge('A', 'D');
graph.addEdge('A', 'E');
graph.addEdge('B', 'C');
graph.addEdge('C', 'E');
graph.addEdge('C', 'F');
graph.addEdge('D', 'E');
graph.addEdge('E', 'F');

// graph.printGraph();

// graph.bfs('A');
graph.printLevelsUsingBFS('A');
graph.dfs('A');

File: python/ctci/ds/graph.py


class Graph:
    def __init__(self):
        self.graph = {}

    def add_vertex(self, vertex):
        if vertex in self.graph:
            print(f"Vertex {vertex} already exists")
        self.graph[vertex] = []

    def add_edge(self, v1, v2):
        self.graph[v1].append(v2)
        self.graph[v2].append(v1)


if __name__ == "__main__":
    g = Graph()
    for v in ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'O']:
        g.add_vertex(v)

    g.add_edge('A', 'B')
    g.add_edge('A', 'C')
    g.add_edge('B', 'D')
    g.add_edge('C', 'D')
    g.add_edge('C', 'E')
    g.add_edge('E', 'F')
    g.add_edge('F', 'G')
    g.add_edge('F', 'O')
    g.add_edge('F', 'I')
    g.add_edge('G', 'H')
    g.add_edge('I', 'J')
    g.add_edge('J', 'K')
    g.add_edge('J', 'L')

    print(g.graph)
